core/src/main/java/org/apache/iceberg/actions/SizeBasedDataRewriter.java

```diff
@@ -84,13 +87,34 @@ private boolean shouldRewrite(List<FileScanTask> group) {
     return enoughInputFiles(group)
         || enoughContent(group)
         || tooMuchContent(group)
-        || anyTaskHasTooManyDeletes(group);
+        || anyTaskHasTooManyDeletes(group)
+        || anyTaskHasTooHighDeleteRatio(group);
   }
 
   private boolean anyTaskHasTooManyDeletes(List<FileScanTask> group) {
     return group.stream().anyMatch(this::tooManyDeletes);
   }
 
+  private boolean anyTaskHasTooHighDeleteRatio(List<FileScanTask> group) {
+    return group.stream().anyMatch(this::tooHighDeleteRatio);
+  }
+
+  private boolean tooHighDeleteRatio(FileScanTask task) {
+    if (null == task.deletes() || task.deletes().isEmpty()) {
+      return false;
+    }
+
+    if (ContentFileUtil.containsSingleDV(task.deletes())
+        || task.deletes().stream().allMatch(ContentFileUtil::isFileScoped)) {
+      long sum = task.deletes().stream().mapToLong(ContentFile::recordCount).sum();
+      double deletedRecords = (double) Math.min(sum, task.file().recordCount());
+      double deleteRatio = deletedRecords / task.file().recordCount();
+      return deleteRatio >= 0.3;
```

Comment:
Can we define a private constant like `DELETE_RATIO_ THRESHOLD` or something? No need to make it configurable in this PR but we probably should eventually.